### ISSUE TITLE
gRPC Truncated/Corrupted Path

### ISSUE THREAD

**Issue by bclermont** [containous/traefik#3385]
Traefik 1.6.2

I'm using` gRPC Client --https--> [Traefik] --http--> [gRPC server]`

it's almost working, but resource path break at some point:

Traefik log:

```
time="2018-05-25T09:34:58Z" level=debug msg="vulcand/oxy/roundrobin/rr: begin ServeHttp on request" Request="{\"Method\":\"POST\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/grpc.health.v1.Health/Check\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\"},\"Proto\":\"HTTP/2.0\",\"ProtoMajor\":2,\"ProtoMinor\":0,\"Header\":{\"Content-Type\":[\"application/grpc\"],\"Te\":[\"trailers\"],\"User-Agent\":[\"grpc-go/1.12.0\"]},\"ContentLength\":-1,\"TransferEncoding\":null,\"Host\":\"account.robotinfra.tools:443\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"14.192.208.43:13516\",\"RequestURI\":\"/grpc.health.v1.Health/Check\",\"TLS\":null}"
```

Path is `/grpc.health.v1.Health/Check`

Log from gRPC Server:

```
zap/grpclogger.go:45	grpc: Server.Serve failed to create ServerTransport: connection error: desc = "transport: http2Server.HandleStreams received bogus greeting from client: \"POST /grpc.health.v1.Hea\""	{"system": "grpc", "grpc_log": true}
```

Path is now `/grpc.health.v1.Hea`

If I try to `GET` `/grpc.health.v1.Health/Check`:

```
time="2018-05-25T09:40:42Z" level=debug msg="vulcand/oxy/forward/http: begin ServeHttp on request" Request="{\"Method\":\"GET\",\"URL\":{\"Scheme\":\"http\",\"Opaque\":\"\",\"User\":null,\"Host\":\"172.18.0.3:8000\",\"Path\":\"\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\"},\"Proto\":\"HTTP/2.0\",\"ProtoMajor\":2,\"ProtoMinor\":0,\"Header\":{\"Accept\":[\"text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8\"],\"Accept-Encoding\":[\"gzip, deflate, br\"],\"Accept-Language\":[\"en-GB,en;q=0.5\"],\"Cache-Control\":[\"max-age=0\"],\"Upgrade-Insecure-Requests\":[\"1\"],\"User-Agent\":[\"Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0\"]},\"ContentLength\":0,\"TransferEncoding\":null,\"Host\":\"account.robotinfra.tools\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"14.192.208.43:10955\",\"RequestURI\":\"/grpc.health.v1.Health/Check\",\"TLS\":null}"
```

```
zap/grpclogger.go:45	grpc: Server.Serve failed to create ServerTransport: connection error: desc = "transport: http2Server.HandleStreams received bogus greeting from client: \"GET /grpc.health.v1.Heal\""	{"system": "grpc", "grpc_log": true}
```

now there is an extra character `l`

like if because I `GET` there is an extra char available (`len("GET") < len("POST")`)

If I cut down the path to less characters and I `GET` path `/grpc.health.v1.H`

```
time="2018-05-25T09:42:17Z" level=debug msg="vulcand/oxy/roundrobin/rr: competed ServeHttp on request" Request="{\"Method\":\"GET\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/grpc.health.v1.H\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\"},\"Proto\":\"HTTP/2.0\",\"ProtoMajor\":2,\"ProtoMinor\":0,\"Header\":{\"Accept\":[\"text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8\"],\"Accept-Encoding\":[\"gzip, deflate, br\"],\"Accept-Language\":[\"en-GB,en;q=0.5\"],\"Upgrade-Insecure-Requests\":[\"1\"],\"User-Agent\":[\"Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0\"]},\"ContentLength\":0,\"TransferEncoding\":null,\"Host\":\"account.robotinfra.tools\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"14.192.208.43:10955\",\"RequestURI\":\"/grpc.health.v1.H\",\"TLS\":null}"
```

path look good `/grpc.health.v1.H`

and then

```
zap/grpclogger.go:45	grpc: Server.Serve failed to create ServerTransport: connection error: desc = "transport: http2Server.HandleStreams received bogus greeting from client: \"GET /grpc.health.v1.H HT\""	{"system": "grpc", "grpc_log": true}
```

`/grpc.health.v1.H HT`

this is even more weird... the path get this weird ` HT` appended to it

**Comment by traefiker** [comment]
Hi! I'm Træfiker :robot: the bot in charge of communication regulation.

Thanks for your interest in Træfik!

Issue templates help us help you by providing all necessary information.

Please edit your issue and use the available templates:

- [Feature request](https://raw.githubusercontent.com/containous/traefik/master/.github/ISSUE_TEMPLATE/Feature_request.md)
- [Bug report](https://raw.githubusercontent.com/containous/traefik/master/.github/ISSUE_TEMPLATE/Bug_report.md)

And remember: each time someone ignores the template, a cute little bunny dies.
<details>
<summary>#SaveTheCuteBunny :heart: :rabbit: :heart:</summary>

![southpark - estrella kill rabbit](https://user-images.githubusercontent.com/5674651/27143914-3cea7896-5130-11e7-81f8-02d8ac2da7a7.gif)

![southpark - estrella kill rabbit2](https://user-images.githubusercontent.com/5674651/27143924-49104c72-5130-11e7-82ae-ccedbd5dfa1c.gif)

</details>

**Comment by Juliens** [comment]
Thanks for your interest in Træfik!

With 1.6.2, you can't do:
`gRPC Client --https--> [Traefik] --http--> [gRPC server]`
You have to do (https with grpc server):
`gRPC Client --https--> [Traefik] --https--> [gRPC server]`

It is explained in the documentation https://docs.traefik.io/user-guide/grpc/.

If you don't have https, you don't have HTTP2 and that what the logs says.

But,

Thanks to this #3371 you will be able to it (https://docs.traefik.io/master/user-guide/grpc/). This feature is already merged in master and will be available in 1.7.
